soc/nordic/nrf54h/power.c

```diff
@@ -0,0 +1,159 @@
+#include <power.h>
+#include "pm_s2ram.h"
+
+static void suspend_common(void)
+{
+
+	/* Flush, disable and power down DCACHE */
+	sys_cache_data_flush_all();
+	sys_cache_data_disable();
+	nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+					RAMBLOCK_CONTROL_BIT_DCACHE, false);
+
+	if (IS_ENABLED(CONFIG_ICACHE)) {
+		/* Disable and power down ICACHE */
+		sys_cache_instr_disable();
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+						RAMBLOCK_CONTROL_BIT_ICACHE, false);
+	}
+
+	/* Disable retention */
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+}
+
+void nrf_poweroff(void)
+{
+	nrf_resetinfo_resetreas_local_set(NRF_RESETINFO, 0);
+	nrf_resetinfo_restore_valid_set(NRF_RESETINFO, false);
+
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+
+	/* TODO: Move it around k_cpu_idle() implementation. */
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+
+	suspend_common();
+
+	nrf_lrcconf_task_trigger(NRF_LRCCONF010, NRF_LRCCONF_TASK_SYSTEMOFFREADY);
+
+	__set_BASEPRI(0);
+	__ISB();
+	__DSB();
+	__WFI();
+
+	CODE_UNREACHABLE;
+}
+
+__weak void z_pm_soc_resume(void)
+{
+}
+
+/* Resume domain after local suspend to RAM. */
+void z_pm_sys_resume(void)
+{
+	if (IS_ENABLED(CONFIG_ICACHE)) {
+		/* Power up and re-enable ICACHE */
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+							RAMBLOCK_CONTROL_BIT_ICACHE, true);
+		sys_cache_instr_enable();
+	}
+
+	if (IS_ENABLED(CONFIG_DCACHE)) {
+		/* Power up and re-enable DCACHE */
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+							RAMBLOCK_CONTROL_BIT_DCACHE, true);
+		sys_cache_data_enable();
+	}
+
+	/* Re-enable domain retention. */
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, true);
+
+	/* TODO: Move it around k_cpu_idle() implementation. */
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN,
+			!IS_ENABLED(CONFIG_SOC_NRF54H20_CPURAD));
+
+	z_pm_soc_resume();
+}
+
+/* Function called during local domain suspend to RAM. */
+int z_pm_sys_suspend_to_ram(void)
+{
+	/* Set intormation which is used on domain wakeup to determine if resume from RAM shall
+	 * be performed.
+	 */
+	nrf_resetinfo_resetreas_local_set(NRF_RESETINFO,
+					  NRF_RESETINFO_RESETREAS_LOCAL_UNRETAINED_MASK);
+	nrf_resetinfo_restore_valid_set(NRF_RESETINFO, true);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+
+	suspend_common();
+
+	__set_BASEPRI(0);
+	__ISB();
+	__DSB();
+	__WFI();
+	/*
+	 * We might reach this point is k_cpu_idle returns (there is a pre sleep hook that
+	 * can abort sleeping.
+	 */
+	return -EBUSY;
```

Comment:
I think that this is not the same case (https://github.com/zephyrproject-rtos/zephyr/pull/73095#discussion_r1632994837)  - `__set_BASEPRI(0)` doesn't disable interrupts but clears the interrupt mask. @hubertmis , correct me if I'm wrong.

Reply:
How about the modifications to RESETINFO like setting RESTOREVALID, could those have unintended consequences when we don't actually suspend here? I guess the stored CPU context is no longer valid at this point

Reply:
You're right, the CPU context is no longer valid, but we don't check the RESETINFO then - it happens only when reset vector is called after wake-up

Reply:
I think the only edge case that could cause issues would be domain resets initiated from the secure domain, but I think the UNRETAINEDWAKE bit in RESETREAS.LOCAL will always be cleared in those instances so it should be fine.

Reply:
> I think that this is not the same case ([#73095 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/73095#discussion_r1632994837)) - `__set_BASEPRI(0)` doesn't disable interrupts but clears the interrupt mask. @hubertmis , correct me if I'm wrong.

From ARMv8M manual:

> Software can set this field to a priority number between 1 and the maximum supported priority number. This boosts the current execution priority to that number, masking all exceptions with a lower priority.

Setting BASEPRI to 0 effectively locks all IRQs (including ZLI).

> How about the modifications to RESETINFO like setting RESTOREVALID, could those have unintended consequences when we don't actually suspend here? I guess the stored CPU context is no longer valid at this point

When this function returns, it should enter the "resume" procedure which in both cases of wake-up and returning error here should clean up after suspending things like LRCCONF settings or RESETINFO.RESTOREVALID.

Reply:
Ok, I see there is no such "resume" procedure and LRCCONFs are never restored. That's something we need to improve.

Reply:
Or is it `z_pm_sys_resume()`? This code is so all over the place with functions with similar naming it's hard to understand what's going on in which procedure ;)

Reply:
> Or is it `z_pm_sys_resume()`? This code is so all over the place with functions with similar naming it's hard to understand what's going on in which procedure ;)

Yup - it is `z_pm_sys_resume()`

Reply:
> How about the modifications to RESETINFO like setting RESTOREVALID, could those have unintended consequences when we don't actually suspend here? I guess the stored CPU context is no longer valid at this point

I missed that `pm_s2ram_mark_check_and_clear()` should [always be called when this function returns](https://github.com/zephyrproject-rtos/zephyr/blob/51e52eaeab3195fcabcebf8562499533e8b59582/arch/arm/core/cortex_m/pm_s2ram.S#L70-L86), so my particular question about RESETINFO is not valid since it is indeed cleaned up on failure.